realm/src/main/java/io/realm/internal/async/QueryUpdateTask.java

```diff
@@ -0,0 +1,319 @@
+                        long handoverTableViewPointer = TableQuery.nativeFindAllMultiSortedWithHandover(
+                                sharedGroup.getNativePointer(),
+                                sharedGroup.getNativeReplicationPointer(),
+                                queryEntry.handoverQueryPointer,
+                                0, Table.INFINITE, Table.INFINITE,
+                                queryEntry.queryArguments.columnIndices,
+                                queryEntry.queryArguments.ascendings);
+
+                        result.updatedTableViews.put(queryEntry.element, handoverTableViewPointer);
+                        queryEntry.handoverQueryPointer = 0L;
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Query mode " + queryEntry.queryArguments.type + " not supported");
+                }
+            } else {
+                for (Long handoverQueryPointer : result.updatedTableViews.values()) {
+                    if (handoverQueryPointer != 0) {
+                        TableQuery.nativeCloseQueryHandover(handoverQueryPointer);
+                    }
+                }
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean updateRealmObjectQuery(SharedGroup sharedGroup, Result result) {
+        if (!isTaskCancelled()) {
+            switch (realmObjectEntry.queryArguments.type) {
+                case RealmQuery.ArgumentsHolder.TYPE_FIND_FIRST: {
+                    long handoverRowPointer = TableQuery.
+                            nativeFindWithHandover(sharedGroup.getNativePointer(),
+                                    sharedGroup.getNativeReplicationPointer(),
+                                    realmObjectEntry.handoverQueryPointer, 0);
+                    result.updatedRow.put(realmObjectEntry.element, handoverRowPointer);
+                    break;
+                }
+                default:
+                    throw new IllegalArgumentException("Query mode " + realmObjectEntry.queryArguments.type + " not supported");
+            }
+        } else {
+            TableQuery.nativeCloseQueryHandover(realmObjectEntry.handoverQueryPointer);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isTaskCancelled() {
+        // no point continuing if the caller thread was stopped or this thread was interrupted
+        return Thread.currentThread().isInterrupted();
+    }
+
+    private boolean isAliveHandler(Handler handler) {
+        return handler != null && handler.getLooper().getThread().isAlive();
+    }
+
+    // result of the async query
+    public static class Result {
+        public IdentityHashMap<WeakReference<RealmResults<?>>, Long> updatedTableViews;
+        public IdentityHashMap<WeakReference<RealmObject>, Long> updatedRow;
+        public SharedGroup.VersionID versionID;
+
+        public static Result newRealmResults () {
```

Comment:
Perhaps `newRealmResultsInstance()` and `newRealmObjectInstance` .. the current naming seems a bit unclear IMO